aip/0194.md

```diff
@@ -0,0 +1,237 @@
+- `UNIMPLEMENTED`: The RPC may be implemented in the future.
+- `UNKNOWN`: `UNKNOWN` and `INTERNAL` are both generic error codes - they are
+  errors that may be generated from a wide breadth of causes. Known services
+  within the system **should** generate `INTERNAL` errors when the error doesn't
+  fit any other code. `UNKNOWN` errors are reserved for truly
+  unknown-to-the-system errors, and therefore **should** immediately surfaced to
+  the user.
+
+## Appendix
+
+### Idempotency and pragmaticism
+
+A common concern with retrying `CREATE` and `DELETE` operations is that they are
+non-idempotent operations. The functional concern can be summarized with this
+example:
+
+```
+Get("Foo")
+> NOT_FOUND
+Create("Foo") // Let's work under the assumption that this should be retried.
+// ----> UNAVAILABLE
+// ----> Create("Foo") # retry #1
+// ----> UNAVAILABLE   # actually, the request succeeded, but the connection died during the response
+// ----> Create("Foo") # retry #2
+// ----> ALREADY_EXISTS
+> ALREADY_EXISTS
+```
+
+Here, an unexpected result - `ALREADY_EXISTS` - is returned from a Create. The
+resource didn't exist before the call was initiated, so the user may wonder,
+"How did I receive this result?".
+
+Does this unexpected result make the call non-idempotent?
+
+The definition of idempotency is unfortunately vague in this regard. [Oxford
+dictionary defines idempotent as](https://www.lexico.com/en/definition/idempotent),
+
+> denoting an element of a set which is unchanged in value when multiplied or
+> otherwise operated on by itself.
+
+One interpretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _response_.
+
+This definition allows for retrying `READ` and `UPDATE` requests, since
+repeated runs always result in the same state and response code. However, this
+definition precludes retrying `CREATE` calls, since `ALREADY_EXISTS` may be
+returned unexpectedly (as in the above example). It also precludes retrying
+`DELETE` calls, since `NOT_FOUND` may be returned unexpectedly.
+
+Another intepretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _state_ (or, value).
+
+This second definition allows for retrying all CRUD operations, since the
+interpretation is only concerned with state, and repeating any CRUD operation
+always results in the same state.
+
+This document's retryability guidances follows the latter interpretation,
+which allows for retrying all CRUD operations. The reason for this is that
+there's much more value for users to retry more operations, with little downside.
+
+Consider the downside to retrying the `CREATE` and `DELETE` operations. For
+users it generally just results in an extra code check:
+
+```
+resp, err := client.Create("Foo)
+if err != nil {
+  if status.Convert(err).Code() == codes.ALREADY_EXISTS { // <- extra code check
+    // It already exists.
+    resp = client.Get("Foo")
```

Comment:
Hmm... depending on the application logic, there could be a much more detailed check here. Maybe the foo that's been created is a very different one, created by a separate process, and this really does mean "the create I tried to do failed".

Reply:
That case is covered tangentially in the asynchrony fallacy section below. (such a user should not be performing retries at all)